### PR TITLE
Add Read-Only Notebook View

### LINKED ISSUE #28531: Audit logging not able to show Ad-hoc GUI query
**Describe the bug**
Audit logging not able to show ad-hoc GUI query.

**To Reproduce**

1. New -> GUI query > Sample Dataset > Orders -> Visualize 
2. Admin > Audit > Team members > Audit Log > click the top "Ad-hoc"
3. The page refreshes without showing any results. It just shows "Something went wrong" 

image

**Expected behavior**
The GUI query loads

**Screenshots/videos**
If applicable, add screenshots/videos to help explain your problem.

**Information about your Metabase Installation:**

Able to replicate on 1.45.3 and `master`

**Additional context**
Related to https://github.com/metabase/metabase/issues/16845 since it works for Native SQL but only fails for GUI

### LINKED ISSUE #28639: Ad-hoc GUI queries in the audit log detail view will need a different representation
Ad hoc questions because of https://github.com/metabase/metabase-enterprise/issues/313, but ad hoc GUI queries in the audit log had on their detail view an instance of the old GUI query builder so you could see what the query was.

Since version 1.33.3+ includes the new query builders, we'll need to change how ad hoc queries are represented in the audit log, maybe by displaying the full notebook mode representation of the query.

### LINKED ISSUE #29456: Audit log crashes for ad-hoc gui questions (EE)
**Describe the bug**

The audit log crashes when trying to view an ad-hoc GUI question

**To Reproduce**

1. perform an ad-hoc query
2. view the audit log for that query 
3. see page crash 

https://www.loom.com/share/e70a17838e27463e9508e651a8288328

**Expected behavior**
Page should not crash

**Screenshots/videos**
testauditlog

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "18.0.1+0",
    "java.vendor": "Homebrew",
    "java.vendor.url": "https://github.com/Homebrew/homebrew-core/issues",
    "java.version": "18.0.1",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "18.0.1+0",
    "os.name": "Mac OS X",
    "os.version": "12.4",
    "user.language": "en",
    "user.timezone": "America/Denver"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "postgres",
      "mysql"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "2.1.214 (2022-06-13)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "2.1.214 (2022-06-13)"
      }
    },
    "run-mode": "dev",
    "version": {
      "date": "2023-03-20",
      "src_hash": "fc3f1f02456d1cbff0c4f00a8469e13e4d2c3255",
      "tag": "v0.46.1-SNAPSHOT",
      "branch": "backport-29295",
      "hash": "737873c"
    },
    "settings": {
      "report-timezone": "America/Detroit"
    }
  }
}
```

**Severity**

P2/P3 - app crash is nasty but it's a minor feature and has a workaround: exiting admin and hitting the browser back button.

**Additional context**

### PR DESCRIPTION
**Pull request by iethree**
Resolves https://github.com/metabase/metabase/issues/29456
Resolves https://github.com/metabase/metabase/issues/28531
Resolves https://github.com/metabase/metabase/issues/28639

### Description

Until now, the audit log was showing the old/deprecated GUI query builder (instead of the notebook) to graphically represent queries not built with SQL.  Most importantly, this component started throwing page-crashing errors in the audit log. Instead of fixing deprecated code, I opted to replace that view with a simple read-only version of the notebook query builder.

1. Creates a New `QueryViewer` container that pulls in appropriate metadata for a query and shows either the notebook editor or native query editor in read only mode, as appropriate.
2. Creates a new `ReadOnlyNotebook` component which passes down `readOnly` props through the notebook editor to hide various UI elements that suggest you should be clicking on them.
3. Adds unit test coverage, which I think is the first and only unit test coverage for the notebook editor 🥳 

**Before**
(when it didn't crash)
Screen Shot 2023-03-24 at 11 02 56 AM

**After**

Screen Shot 2023-03-23 at 3 23 28 PM

### How to verify

1. Startup metabase enterprise
2. open a new question and create a GUI question with filters, aggregations, limits, etc.
3. open another new question and create a sql question
5. Open the audit log by going to admin settings > audit > audit log
6. click on the two ad-hoc queries you made and see that both notebook and native queries display properly
7. ensure that both query viewers are non-interactive
8. refresh the page on the audit page and ensure that the query displays still display properly there

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29525)

### CODE REVIEW

File: enterprise/frontend/src/metabase-enterprise/audit_app/lib/mode.js [hunk 1]
```diff
@@ -76,8 +76,10 @@ const AuditDrill = ({ question, clicked }) => {
           name: "detail",
           title: `View this`,
           default: true,
-          url() {
-            return `/admin/audit/query/${encodeURIComponent(String(value))}`;
+          action() {
+            return push(
+              `/admin/audit/query/${encodeURIComponent(String(value))}`,
+            );
```
Comment (iethree) [inline enterprise/frontend/src/metabase-enterprise/audit_app/lib/mode.js:82]: i think forcing a page load was a hack to keep this page on life support, it wasn't working, so I changed it to use react-router which greatly improved performance.

File: frontend/src/metabase/query_builder/components/notebook/ReadOnlyNotebook.unit.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+import Question from "metabase-lib/Question";
+
+import ReadOnlyNotebook from "./ReadOnlyNotebook";
+
+const makeQuery = (options: any) => {
+  // we have to cast this because we have 2 incompatible DatasetQuery types
+  return createMockStructuredDatasetQuery({
+    query: options,
+  }) as DatasetQuery;
+};
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/notebook/ReadOnlyNotebook.unit.spec.tsx:25]: this is kinda nasty, but the StructuredQuery type in our api types is missing most of its properties. Adding them, however, throws tons of type errors throughout the app 😬  Definitely worth fixing, but I didn't want to blow out the scope of this PR by opening that can of worms.

File: frontend/src/metabase/query_builder/containers/QueryViewer.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  useEffect(() => {
+    async function loadMetadata() {
+      await dispatch(loadMetadataForCard(card as any) as any);
+    }
+    loadMetadata();
+  }, [card, dispatch]);
+
+  const question = new Question(card, metadata);
+
+  const query = question.query();
```
Comment (iethree) [inline frontend/src/metabase/query_builder/containers/QueryViewer.tsx:33]: I would appreciate any suggestions for improving the use of redux hooks here. We don't really have an established pattern for this yet. 

This works, but certainly isn't very elegant. The `any` typecasting in particular is gross, but our redux actions don't seem to play well with typescript. 😢

**Comment by iethree** [comment]
@paoliniluis just wanted to call this PR to your attention since it addresses a couple issues you filed.

**Review by alxnddr — APPROVED** [review]
LGTM! Would you be able to backport just P1 fixes to x.46 without bringing the new read-only notebook?

**Comment by iethree** [comment]
> LGTM! Would you be able to backport just P1 fixes to x.46 without bringing the new read-only notebook?

No, the P1 crash was in the GUI editor, which I decided not to fix since it's deprecated.

**Pull request merged**